proto/api.go

```diff
@@ -161,41 +312,87 @@ func (rh *ResponseHeader) Combine(otherRH *ResponseHeader) {
 			rh.Txn = nil
 		}
 	}
+	return nil
 }
 
 // Combine implements the Combinable interface.
-func (sr *ScanResponse) Combine(c Response) {
+func (sr *ScanResponse) Combine(c Response) error {
 	otherSR := c.(*ScanResponse)
 	if sr != nil {
 		sr.Rows = append(sr.Rows, otherSR.GetRows()...)
-		sr.Header().Combine(otherSR.Header())
+		if err := sr.Header().Combine(otherSR.Header()); err != nil {
+			return err
+		}
 	}
+	return nil
 }
 
 // Combine implements the Combinable interface.
-func (sr *ReverseScanResponse) Combine(c Response) {
+func (sr *ReverseScanResponse) Combine(c Response) error {
 	otherSR := c.(*ReverseScanResponse)
 	if sr != nil {
 		sr.Rows = append(sr.Rows, otherSR.GetRows()...)
-		sr.Header().Combine(otherSR.Header())
+		if err := sr.Header().Combine(otherSR.Header()); err != nil {
+			return err
+		}
 	}
+	return nil
 }
 
 // Combine implements the Combinable interface.
-func (dr *DeleteRangeResponse) Combine(c Response) {
+func (dr *DeleteRangeResponse) Combine(c Response) error {
 	otherDR := c.(*DeleteRangeResponse)
 	if dr != nil {
 		dr.NumDeleted += otherDR.GetNumDeleted()
-		dr.Header().Combine(otherDR.Header())
+		if err := dr.Header().Combine(otherDR.Header()); err != nil {
+			return err
+		}
 	}
+	return nil
 }
 
 // Combine implements the Combinable interface.
-func (rr *ResolveIntentRangeResponse) Combine(c Response) {
+func (rr *ResolveIntentRangeResponse) Combine(c Response) error {
 	otherRR := c.(*ResolveIntentRangeResponse)
 	if rr != nil {
-		rr.Header().Combine(otherRR.Header())
+		if err := rr.Header().Combine(otherRR.Header()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ResetAll resets all the contained requests to their original state.
+func (br *BatchResponse) ResetAll() {
+	for _, rsp := range br.Responses {
+		// TODO(tschottdorf) `rsp.Reset()` doesn't actually clear it out, I'm
+		// not sure why though.
```

Comment:
It's because `BatchResponse.responses` is not nullable, so `rsp` is a shallow copy of the `ResponseUnion`.